File: IntegrationProject/java/src/com/project/math/Function.java
package com.project.math;
import java.util.ArrayList;
import java.util.List;

/**
 * Function - relation between a set of inputs and a set of
 * outputs. Each input is related to exactly one output. 
 * 
 * @param <E> - coefficients
 * @param <C> - exponents
 */
public abstract class Function<E, C> implements Cloneable {
   protected List<E> exponents;
   protected List<C> coefficients;

   // Constructors
   public Function() {
      exponents = new ArrayList<E>();
      coefficients = new ArrayList<C>();
   }

   public Function(List<E> exponents, List<C> coefficients) {
      this.exponents = exponents;
      this.coefficients = coefficients;
   }

   public Function(Function function) {
      this.exponents = new ArrayList<E>(function.exponents);
      this.coefficients = new ArrayList<C>(function.coefficients);
   }

   // setters
   public void setExponent(List<E> exponents) {
      this.exponents = exponents;
   }

   public void setCoefficient(List<C> coefficients) {
      this.coefficients = coefficients;
   }

   // getters
   public List<E> getExponent() {
      return exponents;
   }

   public List<C> getCoefficient() {
      return coefficients;
   }

   /**
    * addToExponent - Add an exponent to the list
    *
    * @param exponent E
    */
   public void addToExponent(E exponent) {
      exponents.add(exponent);
   }

   /**
    * addToExponent - Add an exponent to the list at provided index
    *
    * @param index - integer of where to put the exponent
    * @param exponent E
    */
   public void addToExponent(int index, E exponent) {
      exponents.add(index, exponent);
   }

   /**
    * addToCoefficient - Add a coefficient to the list
    *
    * @param exponent C
    */
   public void addToCoefficient(C coefficient) {
      coefficients.add(coefficient);
   }

   /**
    * addToCoefficient - Add a coefficient to the list at provided index
    *
    * @param index - integer of where to put the coefficient
    * @param exponent C
    */
   public void addToCoefficient(int index, C coefficient) {
      coefficients.add(index, coefficient);
   }

   /**
    * getCoefficientList() - retrieve the coefficient list as string
    *
    * @return - returns the list of coefficients as a string
    */
   public String getCoefficientList() {
      String coefficientList = "";
      for(int index = 0; index < coefficients.size(); index++) {
         if(coefficientList.isEmpty()) {
            coefficientList += coefficients.get(index);
         }
         else {
            coefficientList += " " + coefficients.get(index);
         }
      }

      return coefficientList;
   }

   /**
    * getExponentList() - retrieve the exponent list as string
    *
    * @return - returns the list of exponents as a string
    */
   public String getExponentList() {
      String exponentList = "";
      for(int index = 0; index < exponents.size(); index++) {
         if(exponentList.isEmpty()) {
            exponentList += exponents.get(index);
         }
         else {
            exponentList += " " + exponents.get(index);
         }
      }

      return exponentList;
   }

   /**
    * clone - hard copy a function instead of shallow copy
    */
   @Override
   public Function clone() throws CloneNotSupportedException {
      return (Function) super.clone();
   }

   // abstract methods that may be different depending on type of function
   public abstract Double evaluate(Double x);

   protected abstract String exponentToString(E exponent);

   protected abstract String coefficientToString(C coefficient);

   public abstract String toString();
}

File: IntegrationProject/cpp/src/Function.cpp
/*
 * Function.cpp
 *
 * Function - relation between a set of inputs and a set of
 * outputs. Each input is related to exactly one output. 
 * 
 * param <E> - coefficients
 * param <C> - exponents
 *
 *  Created on: Dec 26, 2015
 *      Author: chris
 */

#include "Function.h"
#include "Utilities.h"

namespace MathFunctions {
   template <typename E, typename C>
   Function<E, C>::Function() {
   }

   template <typename E, typename C>
   Function<E, C>::Function(const std::vector<E> &exponents, const std::vector<C> &coefficients) {
      this->exponents = exponents;
      this->coefficients = coefficients;
   }

   template <typename E, typename C>
   void Function<E, C>::setCoefficient(const int index, const C coefficient) {
      this->coefficients[index] = coefficient;
   }

   template <typename E, typename C>
   void Function<E, C>::setExponent(const int index, const E exponent) {
      this->exponents[index] = exponent;
   }

   template <typename E, typename C>
   void Function<E, C>::setCoefficients(const std::vector<C> &coefficients) {
      this->coefficients = coefficients;
   }

   template <typename E, typename C>
   void Function<E, C>::setExponents(const std::vector<E> &exponents) {
      this->exponents = exponents;
   }

   template <typename E, typename C>
   C Function<E, C>::getCoefficient(const int index) const {
      return this->coefficients[index];
   }

   template <typename E, typename C>
   E Function<E, C>::getExponent(const int index) const {
      return this->exponents[index];
   }

   template <typename E, typename C>
   std::vector<C> Function<E, C>::getCoefficients() const {
      return this->coefficients;
   }

   template <typename E, typename C>
   std::vector<E> Function<E, C>::getExponents() const {
      return this->exponents;
   }

   template<typename E, typename C>
   E Function<E, C>::getDegree() const {
      return this->exponents[exponents.size() - 1];
   }

   template <typename E, typename C>
   Function<E, C>::~Function() {

   }

   template <typename E, typename C>
   double Function<E, C>::evaluate(double x) {}

   template <typename E, typename C>
   std::ostream &operator <<(std::ostream &out, Function<E, C> &function) {
      // check for basic function of degree 0 
      if(function.getDegree() <= -1) {
         out << 0;
         return out;
      }
      else if(function.getDegree() == 0) {
         out << Utilities::fraction(function.getCoefficient(0), 0.00001);
         return out;
      }
      else {
         // do nothing and continue
      }

      // go up to the first nonzero coefficient
      int indexFirstNonZero = function.getCoefficients().size() - 1;
      while(function.getCoefficient(indexFirstNonZero) == 0) {
         indexFirstNonZero--;
      }

      // first coefficient will be different - is closer to number
      if(function.getCoefficient(indexFirstNonZero) < 0.0 
         && function.getCoefficient(indexFirstNonZero) != -1) {

         out << "-" << Utilities::fraction(std::abs(function.getCoefficient(
            indexFirstNonZero)), 0.00001);
      }
      else if(function.getCoefficient(indexFirstNonZero) > 0.0 
         && function.getCoefficient(indexFirstNonZero) != 1) {

         out << Utilities::fraction(function.getCoefficient(indexFirstNonZero), 0.00001);
      }
      else if(function.getCoefficient(indexFirstNonZero) < 0.0 
         && function.getCoefficient(indexFirstNonZero) == -1) {

         out << "-";
      }
      else if(function.getCoefficient(indexFirstNonZero) > 0.0 
         && function.getCoefficient(indexFirstNonZero) == 1) {

         out << "";
      }
      else {
         // coefficient is 0 do nothing
      }

      // exponent of 0 will not display. if 1 then display x otherwise x^exponent
      if(function.getExponent(indexFirstNonZero) == 0) {
         // exponent 0 do not print x
      }
      else if(function.getExponent(indexFirstNonZero) == 1) {
         out << "x";
      }
      else {
         out << "x^" << Utilities::fraction(function.getExponent(indexFirstNonZero), 0.00001);
      }

      // add other terms to the out buffer
      for(int index = indexFirstNonZero - 1; index >= 0; index--) {
         if(function.getCoefficient(index) != 0) {
            if(function.getCoefficient(index) < 0.0 && function.getCoefficient(index) != -1) {
               out << " - " << Utilities::fraction(std::abs(function.getCoefficient(index)), 0.00001);
            }
            else if(function.getCoefficient(index) > 0.0 && function.getCoefficient(index) != 1) {
               out << " + " << Utilities::fraction(function.getCoefficient(index), 0.00001);
            }
            else if(function.getCoefficient(index) < 0.0 && function.getCoefficient(index) == -1 
               && index != 0) {

               out << " - ";
            }
            else if(function.getCoefficient(index) > 0.0 && function.getCoefficient(index) == 1 
               && index != 0) {

               out << " + ";
            }
            else if(function.getCoefficient(index) < 0.0 && function.getCoefficient(index) == -1 
               && index == 0) {

               out << " - 1";
            }
            else {
               out << " + 1";
            }

            if(function.getExponent(index) == 0) {
               // exponent 0 do not print x
            }
            else if(function.getExponent(index) == 1 && function.getCoefficient(index) != 0) {
               out << "x";
            }
            else {
               out << "x^" << Utilities::fraction(function.getExponent(index), 0.00001);
            }
         }
      }

      return out;
   }
}
